### PR TITLE
ACPI changes

### PR DESCRIPTION
**Pull request by sameo**
This is a PR created for reviewing convenience. It contains our first batch of ACPI patches that we want to send for upstream review. It includes all of our generic ACPI changes and refactoring, together with our hardware-reduced ACPI implementation.

### CODE REVIEW

**Comment by sameo** [comment]
cc @bonzini, in case you're interested. We'll send a version of this PR to the qemu mailing list very shortly.

File: hw/acpi/reduced.c [hunk 1]
```diff
@@ -44,8 +46,30 @@
...
+
+static void acpi_dsdt_add_cpus(MachineState *ms, Aml *dsdt, Aml *scope, int smp_cpus, AcpiConfiguration *conf)
+{
+    CPUHotplugFeatures opts = {
+        .apci_1_compatible = false,
+        .has_legacy_cphp = false,
+    };
+
+    build_cpus_aml(dsdt, ms, opts, conf->cpu_hotplug_io_base,
+                   "\\_SB.PCI1", GED_DEVICE);
```
Comment (liujing2) [inline hw/acpi/reduced.c]: @sameo  Does \\_SB.PCI0 work?

Comment (liujing2) [reply hw/acpi/reduced.c]: @sameo I found later this would be changed to \\_SB, but what I concern is if PCI1 will be a discussion issue in this patch.

Comment (sameo) [reply hw/acpi/reduced.c]: @liujing2 This was a mistake from my part. This now fixed, we never use `PCI1`.

File: include/hw/acpi/reduced.h [hunk 2]
```diff
@@ -0,0 +1,6 @@
+#ifndef HW_ACPI_REDUCED_H
```
Comment (rbradford) [inline include/hw/acpi/reduced.h:19]: Does this file need a copyright header?

File: hw/acpi/reduced.c [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+    GArray *table_offsets;
+    unsigned dsdt, xsdt;
+    Range pci_hole, pci_hole64;
+    Object *pci_host;
+    PCIBus *bus = NULL;
+    GArray *tables_blob = tables->table_data;
+
+    acpi_get_pci_holes(&pci_hole, &pci_hole64);
+    table_offsets = g_array_new(false, true /* clear */,
+                                        sizeof(uint32_t));
```
Comment (rbradford) [inline hw/acpi/reduced.c]: Is this consistent with the coding style?

File: include/hw/boards.h [hunk 4]
```diff
@@ -5,6 +5,7 @@
```
Comment (rbradford) [inline include/hw/boards.h]: Typo of generic in the commit message for 3d240bcd9dbb781fafd5c2d2093a56e5e7dee44f

File: hw/acpi/aml-build.c [hunk 5]
```diff
@@ -1869,51 +1869,51 @@ Aml *build_crs(PCIHostState *host, CrsRangeSet *range_set)
     return crs;
 }
```
Comment (rbradford) [inline hw/acpi/aml-build.c:1961]: Typo of slightly in commit message 861742a39d538b9919967a9bedee8b28a4475e7d

File: hw/acpi/aml-build.c [hunk 6]
```diff
@@ -2455,6 +2456,134 @@ void build_srat_memory(AcpiSratMemoryAffinity *numamem, uint64_t base,
     numamem->range_length = cpu_to_le64(len);
 }
```
Comment (rbradford) [inline hw/acpi/aml-build.c:2660]: I think the commit message for 88762b272b20f5c331263e87ce5d4c4e4d061174 is wrong, i think it should be architecture *independent*?

File: hw/acpi/aml-build.c [hunk 7]
```diff
@@ -2310,6 +2311,199 @@ Aml *build_pci_host_bridge(Aml *table, AcpiPciBus *pci_host)
...
+    Aml *if_ctx;
+    int32_t devfn = PCI_DEVFN(slot, 0);
+
+    if_ctx = aml_if(aml_and(aml_arg(0), aml_int(0x1U << slot), NULL));
+    aml_append(if_ctx, aml_notify(aml_name("S%.02X", devfn), aml_arg(1)));
+    aml_append(method, if_ctx);
+}
+
+void build_append_pci_bus_devices(Aml *parent_scope, PCIBus *bus,
+                                         bool pcihp_bridge_en)
```
Comment (rbradford) [inline hw/acpi/aml-build.c]: Something looks funky with the whitespace here.

**Review by rbradford — COMMENTED** [review]
Please double-check with the QEMU guidelines whether you need to include your Signed-off-by on the patches from other people.

**Comment by bonzini** [comment]
Why are the hardware-reduced tables included, since they are only used by the virt machine type?

File: include/hw/acpi/acpi.h [hunk 8]
```diff
@@ -186,6 +188,40 @@ extern int acpi_enabled;
...
+    int legacy_acpi_table_size;
+    bool legacy_cpu_hotplug;
+    bool rsdp_in_ram;
+    unsigned acpi_data_size;
+    bool linuxboot_dma_enabled;
+
+    /* Machine state settings */
+    FWCfgState *fw_cfg;
+    HotplugHandler *acpi_dev;
+    ram_addr_t below_4g_mem_size, above_4g_mem_size;;
```
Comment (mcastelino) [inline include/hw/acpi/acpi.h:223]: Are these fields specific to ACPI
```
    ram_addr_t below_4g_mem_size, above_4g_mem_size;;
    uint64_t numa_nodes;
    uint64_t *node_mem;
    bool apic_xrupt_override;
    unsigned apic_id_limit;
```
or just moved here for convenience.

File: hw/i386/acpi-build.c [hunk 9]
```diff
@@ -2803,72 +2795,61 @@ static void acpi_build_reset(void *build_opaque)
...
+                        name, acpi_build_update, conf, NULL, true);
 }
 
 static const VMStateDescription vmstate_acpi_build = {
     .name = "acpi_build",
     .version_id = 1,
     .minimum_version_id = 1,
     .fields = (VMStateField[]) {
-        VMSTATE_UINT8(patched, AcpiBuildState),
+        VMSTATE_BOOL(patched, AcpiBuildState),
```
Comment (mcastelino) [inline hw/i386/acpi-build.c:1824]: Why was this change needed?

File: include/hw/acpi/acpi.h (hunk 8 above)
Comment (mcastelino) [reply include/hw/acpi/acpi.h:223]: Also it would be helpful for the commit message itself to be expanded to explain what got moved around and why.

File: hw/acpi/aml-build.c [hunk 10]
```diff
@@ -1651,6 +1651,69 @@ build_xsdt(GArray *table_data, BIOSLinker *linker, GArray *table_offsets,
                  (void *)xsdt, "XSDT", xsdt_len, 1, oem_id, oem_table_id);
 }
 
+/* Legacy RSDP pointing at an RSDT. This is deprecated */
```
Comment (mcastelino) [inline hw/acpi/aml-build.c:2589]: Would be good to expand the commit message to explain why the code motion was needed. Or maybe split the code motion into a separate commit.

File: hw/acpi/aml-build.c [hunk 11]
```diff
@@ -1651,6 +1651,69 @@ build_xsdt(GArray *table_data, BIOSLinker *linker, GArray *table_offsets,
...
+    /* Checksum to be filled by Guest linker */
+    bios_linker_loader_add_checksum(linker, ACPI_BUILD_RSDP_FILE,
+        (char *)rsdp - rsdp_table->data, sizeof *rsdp,
+        (char *)&rsdp->checksum - rsdp_table->data);
+
+    return rsdp_table;
+}
+
+/* RSDP pointing at an XSDT */
+GArray *build_rsdp(GArray *rsdp_table, BIOSLinker *linker, unsigned xsdt_tbl_offset)
```
Comment (mcastelino) [inline hw/acpi/aml-build.c]: Was build_rsdp modified in any way or just moved around?

File: hw/arm/virt-acpi-build.c [hunk 12]
```diff
@@ -35,6 +35,7 @@
 #include "target/arm/cpu.h"
 #include "hw/acpi/acpi-defs.h"
 #include "hw/acpi/acpi.h"
+#include "hw/acpi/aml-build.h"
```
Comment (mcastelino) [inline hw/arm/virt-acpi-build.c:38]: So you are replacing the implementation here with the new implementation that was added earlier. Maybe expand the commit message to indicate the same.

File: hw/acpi/aml-build.c [hunk 13]
```diff
@@ -1869,51 +1869,51 @@ Aml *build_crs(PCIHostState *host, CrsRangeSet *range_set)
     return crs;
 }
 
-Aml *build_osc_method(void)
+Aml *build_osc_method(uint32_t value)
```
Comment (mcastelino) [inline hw/acpi/aml-build.c]: Could use a comment/header on what value was being used for.
I assume this is the `_OSC Capabilities Buffer Control Field` mask.

File: hw/i386/acpi-build.c [hunk 14]
```diff
@@ -1335,7 +1335,9 @@ build_dsdt(GArray *table_data, BIOSLinker *linker,
         aml_append(dev, aml_name_decl("_CID", aml_eisaid("PNP0A03")));
         aml_append(dev, aml_name_decl("_ADR", aml_int(0)));
         aml_append(dev, aml_name_decl("_UID", aml_int(1)));
-        aml_append(dev, build_osc_method());
+        aml_append(dev, aml_name_decl("SUPP", aml_int(0)));
+        aml_append(dev, aml_name_decl("CTRL", aml_int(0)));
+        aml_append(dev, build_osc_method(0x1F));
```
Comment (mcastelino) [inline hw/i386/acpi-build.c]: I assume this is the `_OSC Control Field` mask.
It may be good to not make it a manifest constant. Instead an ORing of defined constants.

I assume this indicates the following are supported
```
0 PCI Express Native Hot Plug control
1 SHPC Native Hot Plug control
2 PCI Express Native Power Management Events control
3 PCI Express Advanced Error Reporting control
4 PCI Express Capability Structure control
```
and
```
5 Latency Tolerance Reporting control
6 are ignored
```

/cc @sboeuf

File: hw/acpi/aml-build.c [hunk 15]
```diff
@@ -1869,51 +1869,51 @@ Aml *build_crs(PCIHostState *host, CrsRangeSet *range_set)
...
-    aml_append(method, aml_create_dword_field(aml_arg(3), aml_int(0), "CDW1"));
-
-    if_ctx = aml_if(aml_equal(
-        aml_arg(0), aml_touuid("33DB4D5B-1FF7-401C-9657-7441C03DD766")));
-    aml_append(if_ctx, aml_create_dword_field(aml_arg(3), aml_int(4), "CDW2"));
-    aml_append(if_ctx, aml_create_dword_field(aml_arg(3), aml_int(8), "CDW3"));
-
-    aml_append(if_ctx, aml_store(aml_name("CDW3"), a_ctrl));
-
-    /*
```
Comment (mcastelino) [inline hw/acpi/aml-build.c]: We lost these comments which provided some context.

File: hw/acpi/aml-build.c [hunk 16]
```diff
@@ -29,6 +29,25 @@
...
+#define PCI_HOST_BRIDGE_IO_0_MIN_ADDR      0x0000
+#define PCI_HOST_BRIDGE_IO_0_MAX_ADDR      0x0cf7
+#define PCI_HOST_BRIDGE_IO_1_MIN_ADDR      0x0d00
+#define PCI_HOST_BRIDGE_IO_1_MAX_ADDR      0xffff
+#define PCI_VGA_MEM_BASE_ADDR              0x000a0000
+#define PCI_VGA_MEM_MAX_ADDR               0x000bffff
+#define IO_0_LEN                           0xcf8
+#define VGA_MEM_LEN                        0x20000
+
+static const char *pci_hosts[] = {
```
Comment (mcastelino) [inline hw/acpi/aml-build.c:50]: Oh so we moved it around again. This confused me for a bit.

File: hw/acpi/aml-build.c [hunk 17]
```diff
@@ -2095,6 +2159,150 @@ Aml *build_prt(bool is_pci0_prt)
     return method;
 }
 
+Aml *build_pci_host_bridge(Aml *table, AcpiPciBus *pci_host)
```
Comment (mcastelino) [inline hw/acpi/aml-build.c:2192]: We seem a little inconsistent with how we refer to AcpiPciBus, we call it  pci_host in some cases and differently in other places. Let me see if I can find that.

File: hw/acpi/cpu.c [hunk 18]
```diff
@@ -569,9 +569,11 @@ void build_cpus_aml(Aml *table, MachineState *machine, CPUHotplugFeatures opts,
     aml_append(sb_scope, cpus_dev);
     aml_append(table, sb_scope);
 
-    method = aml_method(event_handler_method, 0, AML_NOTSERIALIZED);
-    aml_append(method, aml_call0("\\_SB.CPUS." CPU_SCAN_METHOD));
-    aml_append(table, method);
+    if (event_handler_method) {
+        method = aml_method(event_handler_method, 0, AML_NOTSERIALIZED);
+        aml_append(method, aml_call0("\\_SB.CPUS." CPU_SCAN_METHOD));
```
Comment (mcastelino) [inline hw/acpi/cpu.c:574]: Wish we did something like `MEMORY_DEVICES_CONTAINER` out here too. maybe `CPU_DEVICES_CONTAINER`

File: hw/acpi/aml-build.c [hunk 19]
```diff
@@ -2455,6 +2456,134 @@ void build_srat_memory(AcpiSratMemoryAffinity *numamem, uint64_t base,
...
+    AcpiSystemResourceAffinityTable *srat;
+    AcpiSratMemoryAffinity *numamem;
+
+    int i;
+    int srat_start, numa_start, slots;
+    uint64_t mem_len, mem_base, next_base;
+    MachineClass *mc = MACHINE_GET_CLASS(machine);
+    const CPUArchIdList *apic_ids = mc->possible_cpu_arch_ids(machine);
+    ram_addr_t hotplugabble_address_space_size =
+        object_property_get_int(OBJECT(machine), PC_MACHINE_DEVMEM_REGION_SIZE,
```
Comment (mcastelino) [inline hw/acpi/aml-build.c]: PC_MACHINE_DEVMEM_REGION_SIZE does not seem to be architecture independent.

File: hw/acpi/reduced.c [hunk 20]
```diff
@@ -30,6 +30,8 @@
 #include "hw/acpi/acpi.h"
 #include "hw/acpi/aml-build.h"
 #include "hw/acpi/bios-linker-loader.h"
+#include "hw/acpi/cpu.h"
```
Comment (mcastelino) [inline hw/acpi/reduced.c:34]: This commit could have been part of different patch set.

File: hw/acpi/reduced.c [hunk 21]
```diff
@@ -249,3 +275,136 @@ void acpi_reduced_setup(MachineState *machine, AcpiConfiguration *conf)
      */
     acpi_build_tables_cleanup(&tables, false);
 }
+
+#define CPU_SCAN_METHOD   "CSCN"
```
Comment (mcastelino) [inline hw/acpi/reduced.c]: @sboeuf is this what you are calling new vs legacy?

Comment (sboeuf) [reply hw/acpi/reduced.c]: Yes, this way we should not have to define this again in `reduced.c`.

**Comment by mcastelino** [comment]
Can we split this into two sets of patches. Keeping the generic ACPI changes separate from the hardware reduced ACPI patches. May make it easier to review upstream too. Also we could possibly expand some of the commit messages with more context.

File: hw/acpi/reduced.c (hunk 3 above)
Comment (liujing2) [reply hw/acpi/reduced.c]: Is it good to use ./scripts/checkpatch.pl to check if our patches match the coding style? I also see many lines more of 72 characters.

**Comment by sameo** [comment]
@bonzini 
> Why are the hardware-reduced tables included, since they are only used by the virt machine type?

It's shows why we are doing the ACPI changes. Would you prefer to have the hardware-reduced tables sent as a standalone, separate patch?

File: hw/i386/acpi-build.c (hunk 9 above)
Comment (sameo) [reply hw/i386/acpi-build.c:1824]: Both the arm and i386 machine type had a `AcpiBuildState` and the only difference was the `patched` type. i386 was using a uint8 while arm was using a boolean. When factorizing those, it made more sense to keep it as boolean.

File: include/hw/acpi/reduced.h (hunk 2 above)
Comment (sameo) [reply include/hw/acpi/reduced.h:19]: Fixed.

File: hw/acpi/reduced.c (hunk 3 above)
Comment (sameo) [reply hw/acpi/reduced.c]: All patches are `checkpatch` clean now, thanks for the reminder.

File: include/hw/boards.h (hunk 4 above)
Comment (sameo) [reply include/hw/boards.h]: Fixed.

File: hw/acpi/aml-build.c (hunk 5 above)
Comment (sameo) [reply hw/acpi/aml-build.c:1961]: Fixed.

File: hw/acpi/aml-build.c (hunk 6 above)
Comment (sameo) [reply hw/acpi/aml-build.c:2660]: Yes, fixed. Thanks.

File: hw/acpi/aml-build.c (hunk 7 above)
Comment (sameo) [reply hw/acpi/aml-build.c]: Fixed as well.

File: include/hw/acpi/acpi.h (hunk 8 above)
Comment (sameo) [reply include/hw/acpi/acpi.h:223]: > Are these fields specific to ACPI
> 
> ```
>     ram_addr_t below_4g_mem_size, above_4g_mem_size;;
>     uint64_t numa_nodes;
>     uint64_t *node_mem;
>     bool apic_xrupt_override;
>     unsigned apic_id_limit;
> ```
> or just moved here for convenience.

They're used for building ACPI tables, so not ACPI specific fields. I'll fix the structure to make it clearer.

Comment (sameo) [reply include/hw/acpi/acpi.h:223]: > Also it would be helpful for the commit message itself to be expanded to explain what got moved around and why.

Done.

File: hw/acpi/aml-build.c (hunk 10 above)
Comment (sameo) [reply hw/acpi/aml-build.c:2589]: I split the patch in two, thanks for the suggestion.

File: hw/i386/acpi-build.c (hunk 14 above)
Comment (sameo) [reply hw/i386/acpi-build.c]: Done

File: hw/acpi/aml-build.c (hunk 13 above)
Comment (sameo) [reply hw/acpi/aml-build.c]: Done

File: hw/arm/virt-acpi-build.c (hunk 12 above)
Comment (sameo) [reply hw/arm/virt-acpi-build.c:38]: Done.

File: hw/acpi/aml-build.c (hunk 10 above)
Comment (sameo) [reply hw/acpi/aml-build.c:2589]: I split the patch in two, thanks for the suggestion.

File: hw/acpi/aml-build.c (hunk 17 above)
Comment (sameo) [reply hw/acpi/aml-build.c:2192]: We call it acpi_pci_host in `reduced.c:acpi_reduced_build()`, because there already is a `pci_host` pointer.

File: hw/acpi/aml-build.c (hunk 19 above)
Comment (sameo) [reply hw/acpi/aml-build.c]: Fixed as well.

**Comment by sameo** [comment]
All comments addressed. I also added all the relevant maintainers to the commit messages through Cc.
As far as I'm concerned, this is ready to be sent upstream. @rbradford @mcastelino @sboeuf @liujing2 @yangzhon ?

**Comment by sameo** [comment]
Patches sent upstream for review:
https://lists.gnu.org/archive/html/qemu-devel/2018-10/msg04953.html

**Comment by sameo** [comment]
I reduced the scope of that initial PR and removed the hardware-reduced implementation from it.
Our next patchset will include the hardware-reduced implementation and the virt/arm conversion to it, as requested by @bonzini .

See https://lists.gnu.org/archive/html/qemu-devel/2018-10/msg06209.html

**Comment by sameo** [comment]
> I reduced the scope of that initial PR and removed the hardware-reduced implementation from it.
> Our next patchset will include the hardware-reduced implementation and the virt/arm conversion to it, as requested by @bonzini .
> 
> See https://lists.gnu.org/archive/html/qemu-devel/2018-10/msg06209.html

v3 with the right maintainers cc'ed:

https://lists.gnu.org/archive/html/qemu-devel/2018-10/msg06240.html

**Comment by sameo** [comment]
v4: https://lists.gnu.org/archive/html/qemu-devel/2018-11/msg00055.html

```
v3 -> v4:
   * Renamed all AcpiConfiguration pointers from conf to acpi_conf.
   * Removed the ACPI_BUILD_ALIGN_SIZE export.
   * Temporarily updated the arm virt build_rsdp() prototype for
     bisectability purposes.
   * Removed unneeded pci headers from acpi-build.c.
   * Refactor the acpi PCI host getter so that it truly is architecture
     agnostic, by carrying the PCI host pointer through the
     AcpiConfiguration structure.
   * Splitted the PCI host AML builder API export patch from the PCI host
     and holes getter one.
   * Reduced the build_srat() export scope to hw/i386 instead of the broader
     hw/acpi. SRAT builders are truly architecture specific and can hardly be
     generalized.
   * Completed the ACPI builder documentation.
```

**Comment by sameo** [comment]
v5: https://lists.gnu.org/archive/html/qemu-devel/2018-11/msg00500.html

```
v4 -> v5:
   * Reorganize the ACPI RSDP export and XSDT implementation into 3
     patches.
   * Fix the hw/i386/acpi header inclusions.
```

**Comment by rbradford** [comment]
This project is no longer active and being archived. Cloud Hypervisor is the successor: https://github.com/cloud-hypervisor/cloud-hypervisor
